pathways branch

SCiarella opened the pull request:
fix #144

stefsmeets commented:
Nice work! Let me know when this is finished, and I'll happily review this code.

Couple of things so far:
- Please put any documenting notebooks [here](https://github.com/GEMDAT-repos/gemdat-notebooks) to avoid ballooning this repo.
- Do we need a custom Dijksta implementation? Why not use a well-tested one from scipy or networkx?
- Can you add some tests for these new functions?

SCiarella commented:
The branch has been updated and it is ready for review. In particular, the specific comments by @stefsmeets have been addressed in the following way:

> - Please put any documenting notebooks here to avoid ballooning this repo.

[Done](https://github.com/GEMDAT-repos/gemdat-notebooks/blob/main/paths.ipynb)

> - Do we need a custom Dijksta implementation? Why not use a well-tested one from scipy or networkx?

Switched to networkx

> - Can you add some tests for these new functions?

I am not an expert with testing, so I would like to discuss next time about the 'best practices' in this direction

> - What is this direct path for?

It is used to show that it does not work, as mentioned by @tfamprikis in #144. I have now moved `direct_path` from the main gemdat package to `path.ipynb`, since its use is only pedagogical.

v1kko commented on src/gemdat/transitions.py:
```diff
@@ -401,3 +402,210 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+        path_cost = np.sum(path_energy)
+
+        # Remove all the peaks that have been part of the path
+        peaks_mask[counter] = True
+        for mask in range(counter + 1, len(peaks)):
+            if not peaks_mask[mask] and tuple(peaks[mask]) in path:
+                peaks_mask[mask] = True
+                remaining_peaks -= 1
```
Is this correct? Can a peak not be part of a different path with a different starting point and have a total lower energy?

SCiarella replied:
You are right! I have removed this mask.

v1kko reviewed (approved):
Looks good to me if you add a test for the perculating pathway

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+                     (0, -1, -1), (0, 1, -1), (0, -1, 1), (1, 1, 1),
+                     (-1, -1, -1), (1, -1, -1), (-1, 1, 1)]
+
+    G = nx.Graph()
+    for i in range(F.shape[0]):
+        for j in range(F.shape[1]):
+            for k in range(F.shape[2]):
+                if F[i, j, k] >= 0 and F[i, j, k] < max_energy_threshold:
+                    G.add_node((i, j, k), energy=F[i, j, k])
```
Use ndenumerate: https://numpy.org/doc/stable/reference/generated/numpy.ndenumerate.html

```suggestion

    G = nx.Graph()
    for index, val in np.ndenumerate(F):
        if 0 <= energy < max_energy_threshold:
             G.add_node(index, energy=energy)
```

SCiarella replied:
nice, thanks

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+            for k in range(F.shape[2]):
+                if F[i, j, k] >= 0 and F[i, j, k] < max_energy_threshold:
+                    G.add_node((i, j, k), energy=F[i, j, k])
+    for i, j, k in G.nodes:
+        for move in movements:
+            neighbor = _wrap_pbc(np.array((i, j, k)) + np.array(move), F.shape)
+            if neighbor in G.nodes:
+                G.add_edge((i, j, k), neighbor, weight=F[neighbor])
```
```suggestion
    for node in G.nodes:
        for move in movements:
            neighbor = _wrap_pbc(node + np.array(move), F.shape)
            if neighbor in G.nodes:
                G.add_edge(node, neighbor, weight=F[neighbor])
```

SCiarella replied:
done, thanks

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
     return occupancy
+
+
+def _wrap_pbc(coord: np.ndarray, size: tuple) -> tuple:
```
I think this should return a numpy array.

SCiarella replied:
To be a node or an edge of a graph object it needs to be hashable, but I can make it a tuple outside and leave this with a more general np.ndarray

stefsmeets replied:
Yeah, that makes sense. I saw a bunch of type conversions from array<>tuple, which is what triggered this comment. Either way is okay.

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+    return G
+
+
+def optimal_path(
+    F_graph: nx.Graph,
+    start: tuple,
+    end: tuple,
+) -> typing.Union[list, list]:
+    """Calculate the shortest cost-effective path from start to end using
+    Networkx library.
+
+    Parameters
+    ----------
+    F_graph : nx.Graph
+        Graph of the free energy
+    start : tuple
+        Coordinates of the starting point
+    end: tuple
+        Coordinates of the ending point
+
+    Returns
+    -------
+    path: list[tuple]
+        List of coordinates of the path
+    path_energy: list[float]
+        Energy along the path
+    """
+
+    optimal_path = nx.shortest_path(F_graph,
+                                    source=start,
+                                    target=end,
+                                    weight='weight')
+    path_energy = [F_graph.nodes[node]['energy'] for node in optimal_path]
+
+    return [optimal_path, path_energy]
```
Does this need its own function?

SCiarella replied:
yes because it is also useful for other tasks, like issue #92

stefsmeets replied:
The issue is that this function returns a tuple, which means this function does 2 things, which means I must deal with it in my code if I'm only interested in the path. This is poor separation of concerns. It's not obvious to me why it should return the energy as well. If you must return multiple values, it's usually easier in the long term to return an object (see other comment as well), and hide the energy list behind a method or property. Or split in two functions.

Personally, since this function is so trivial, I think it's easier to write it out in the function calling it.

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+            best_perc_path = path
+            best_perc_path_energy = path_energy
+
+    return total_energy_cost, best_starting_point, best_perc_path, best_perc_path_energy
```
Try to avoid returning a tuple. It's often easier to return an object, like a namedtuple or dataclass and access these variables via an attribute.

SCiarella replied:
Created a path class

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+def find_best_perc_path(
+        F: np.ndarray,
+        peaks: np.ndarray,
+        percolate_x=True,
+        percolate_y=False,
+        percolate_z=False) -> typing.Union[float, tuple, tuple, np.ndarray]:
```
Type hints are missing.

SCiarella replied:
added

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+    X_real, Y_real, Z_real = F.shape
+
+    # Find percolation using virtual images along the required dimensions
+    if not (percolate_x + percolate_y + percolate_z):
```
```suggestion
    if not any(percolate_x, percolate_y, percolate_z):
```

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+    if not (percolate_x + percolate_y + percolate_z):
+        print('Warning: percolation is not defined')
+        return None, None, None
+    if percolate_x:
+        extended_F = np.empty((2 * F.shape[0], F.shape[1], F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:X_real, :, :] = F
+        extended_F[X_real:, :, :] = F
+        F = extended_F
+    if percolate_y:
+        extended_F = np.empty((F.shape[0], 2 * F.shape[1], F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:, :Y_real, :] = F
+        extended_F[:, Y_real:, :] = F
+        F = extended_F
+    if percolate_z:
+        extended_F = np.empty((F.shape[0], F.shape[1], 2 * F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:, :, :Z_real] = F
+        extended_F[:, :, Z_real:] = F
+        F = extended_F
```
I don't understand this code, is this just tiling `F` in 3 directions? Why not use `np.tile`?

SCiarella replied:
Good call! i have just replaced the full block with a single tile call

stefsmeets commented on src/gemdat/volume.py:
```diff
@@ -288,6 +288,26 @@ def to_structure(
+        Parameters
+        ----------
+        kBT : float
+            The temperature of the thermostat
```
What is thermostat?

SCiarella replied:
It is more understandable if we say 'temperature of the simulation', so I have made the change.

However, the temperature fluctuates along the trajectory and the value that we use is the one that the simulation is trying to impose (the thermostat).

stefsmeets commented on src/gemdat/volume.py:
```diff
@@ -288,6 +288,26 @@ def to_structure(
 
+    def get_free_energy(
+        self,
+        kBT,
```
```suggestion
        temperature: float,
```

SCiarella replied:
Agree

stefsmeets commented on src/gemdat/volume.py:
```diff
@@ -288,6 +288,26 @@ def to_structure(
+        """
+        prob = self.data / self.data.sum()
+        free_energy = -kBT * np.log(prob)
+        return np.nan_to_num(free_energy)
```
Maybe not for this PR, but for the long term we could consider that all methods like this return an instance of `Volume` with some flag that tells that it is a probability or free energy, and have some `.units` attribute to keep track of the units.

stefsmeets commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+    for starting_point in peaks:
+
+        # Get the end point which is a periodic image of the peak
+        end_point = (starting_point[0] + X_real * percolate_x,
+                     starting_point[1] + Y_real * percolate_y,
+                     starting_point[2] + Z_real * percolate_z)
```
Define image above loop.

```suggestion
        end_point = starting_point + image
```

SCiarella replied:
done

stefsmeets reviewed (commented):
Hi @SCiarella , Nice work! Just made a pass through the code, some small comments from my side for now. I will have a look at the notebook tomorrow and the PR as a whole! 😃

stefsmeets reviewed (changes requested):
I just checked the notebook, and I'm wondering if you can think a bit of how you want people to use this code. I find some parts not very obvious and I think it can be streamlined quite a bit.

Feel free to tackle this in a follow-up PR if that makes more sense.

Let me know what you think!

## Free energy

This is the current api in this PR:

```python
from gemdat.transitions import optimal_path, free_energy_graph

diff_volume = trajectory_to_volume(diff_trajectory, resolution=0.3)

F = diff_volume.get_free_energy(kBT=trajectory.metadata['temperature'])

F_graph = free_energy_graph(F, max_energy_threshold=1e7, diagonal=True)

start_point = (49, 26, 10)
end_point = (10, 0, 8)

path, path_energy = optimal_path(
    F_graph,
    start_point,
    end_point,
)

fig1 = plots.path_on_landscape(diff_volume, path, structure)
fig1.show()

fig2 = plots.energy_along_path(energy_path=path_energy)
fig2.show()
```

Playing around a little bit, this is what I would suggest we aim for:
- we can grab the temperature from the trajectory metadata
- `F` is a an instance of `Volume` or a subclass of `Volume` with _energy_ related methods
- no need for any extra gemdat imports
- does not hide the networkx graph, so users can do whatever they want with it


```python
import networkx as nx

diff_volume = trajectory.to_volume(resolution=0.3)

F = diff_volume.to_free_energy()
G = F.to_graph(max_energy_threshold=1e7, diagonal=True)

source = (49, 26, 10)
target = (10, 0, 8)

optimal_path = nx.shortest_path(G,
                                source=source,
                                target=target,
                                weight='weight')

fig1 = plots.path_on_landscape(diff_volume, optimal_path, structure)
fig1.show()

energy = [G.nodes[node]['energy'] for node in optimal_path]

fig2 = plots.energy_along_path(energy_path=energy)
fig2.show()
```

## Percolation

Same goes for the percolation, my suggestion would be for an api that looks something like this:
- no need to import anything = more discoverable
- returning a dataclass of some sort makes it easier to extend the code
- hides the wrapping in a method

```python
peaks = F.find_peaks()
best_path = F.find_best_perc_path(peaks, perc_xyz=(True, False, False))

print(f"Total Energy required: {best_path.total_energy_cost}")
print(f"Starting Point: {best_path.starting_point}")
print(f"Best Path: {best_path.best_path}")
print(f"Best Path Energy: {best_path.best_path_energy}")

fig1 = plots.path_on_landscape(diff_volume, structure, path=best_path)
fig1.show()

fig2 = plots.energy_along_path(path=best_path)
fig2.show()
```

SCiarella commented on src/gemdat/plots/plotly/_paths.py:
```diff
@@ -0,0 +1,122 @@
+    fig : matplotlib.figure.Figure
+        Output figure
+    """
+
+    # Create a colormap to visualize the path
+    colormap = plt.get_cmap('viridis')
```
ok

SCiarella commented on src/gemdat/plots/plotly/_paths.py:
```diff
@@ -0,0 +1,122 @@
+
+    fig = density(vol, structure, as_probability=True)
+
+    x_path, y_path, z_path = zip(*path)
+    x_path = np.asarray(x_path) * vol.resolution
+    y_path = np.asarray(y_path) * vol.resolution
+    z_path = np.asarray(z_path) * vol.resolution
```
thanks!

SCiarella commented on src/gemdat/plots/plotly/_paths.py:
```diff
@@ -0,0 +1,122 @@
+            y=y_path,
+            z=z_path,
+            mode='markers+lines',
+            line=dict(width=3),
+            marker=dict(size=6,
+                        color=[
+                            'blue' if i == 0 or i == len(x_path) - 1 else 'red'
+                            for i in range(len(x_path))
+                        ],
+                        symbol='circle',
+                        opacity=0.9),
```
Thanks for the suggestion. Yes, the color is to mark endpoints differently.

SCiarella commented on src/gemdat/plots/plotly/_density.py:
```diff
@@ -158,7 +158,9 @@ def plot_volume(
 
 
-def density(vol: Volume, structure: Optional[Structure] = None) -> go.Figure:
+def density(vol: Volume,
+            structure: Optional[Structure] = None,
+            as_probability: bool = False) -> go.Figure:
```
Created the 'as_probability' method for Volume

SCiarella commented on src/gemdat/plots/plotly/_paths.py:
```diff
@@ -0,0 +1,122 @@
+from ._density import density
+
+
+def energy_along_path(*, energy_path: list) -> plt.Figure:
```
moved

SCiarella commented on src/gemdat/plots/plotly/_paths.py:
```diff
@@ -0,0 +1,122 @@
+    return fig
+
+
+def path_on_grid(*, path: list) -> plt.Figure:
```
moved

SCiarella commented on src/gemdat/plots/plotly/_paths.py:
```diff
@@ -0,0 +1,122 @@
+
+    path_x, path_y, path_z = zip(*path)
+
+    for i in range(len(path) - 1):
+        ax.plot(path_x[i:i + 2],
+                path_y[i:i + 2],
+                path_z[i:i + 2],
```
Yes that is true. I have fixed it.

SCiarella commented on src/gemdat/transitions.py:
```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+    coord: np.ndarray
+        Coordinates of one or multiple sites
+    size: tuple
+        Boudnary size
```
fixed

v1kko commented on src/gemdat/path.py:
```diff
@@ -0,0 +1,211 @@
+    movements = [(1, 0, 0), (-1, 0, 0), (0, 1, 0), (0, -1, 0), (0, 0, 1),
+                 (0, 0, -1)]
+    if diagonal:
+        movements.extend([(1, 1, 0), (-1, -1, 0), (1, -1, 0), (-1, 1, 0),
+                          (1, 0, 1), (-1, 0, -1), (1, 0, -1), (-1, 0, 1),
+                          (0, 1, 1), (0, -1, -1), (0, 1, -1), (0, -1, 1),
+                          (1, 1, 1), (-1, -1, -1), (1, -1, -1), (-1, 1, 1)])
```
```suggestion
        np.append(movements, [(1, 1, 0), (-1, -1, 0), (1, -1, 0), (-1, 1, 0),
                          (1, 0, 1), (-1, 0, -1), (1, 0, -1), (-1, 0, 1),
                          (0, 1, 1), (0, -1, -1), (0, 1, -1), (0, -1, 1),
                          (1, 1, 1), (-1, -1, -1), (1, -1, -1), (-1, 1, 1)])
```

v1kko reviewed (commented):
Looks good to me, I suggested a few small changes, but otherwise I think its good to go, any remaining comments can be converted into issues :rocket:

SCiarella commented on src/gemdat/path.py:
```diff
@@ -0,0 +1,211 @@
+    wrapped_coord: np.ndarray
+        Coordinates inside the PBC
+    """
+    wrapped_coord = coord % size
+    return wrapped_coord
```
Ok, I have removed the function

SCiarella commented on src/gemdat/path.py:
```diff
@@ -0,0 +1,211 @@
+    """
+
+    # Define possible movements in 3D space
+    movements = [(1, 0, 0), (-1, 0, 0), (0, 1, 0), (0, -1, 0), (0, 0, 1),
+                 (0, 0, -1)]
```
changed movements to np.array

SCiarella commented:
I have updated the branch as suggested.  The remaining comments about API and usage can be discussed in another issue

v1kko commented:
> I have updated the branch as suggested. The remaining comments about API and usage can be discussed in another issue

Perfect, after you create the issue, let's merge this :rocket:
